dags/destinations/gads_oci.py

```diff
@@ -0,0 +1,353 @@
+
+  def _send_request(self, customer_id: str, conversions: List[Any]) -> PartialFailures:
+    """Sends conversions to the offline conversion import API.
+
+    Args:
+      customer_id: The customer ID for these conversions.
+      conversions: A list of click conversions.
+
+    Returns: An empty dict if no partial failures exist, or a dict of the index
+      mapped to the error message.
+    """
+    request = self._client.get_type("UploadClickConversionsRequest")
+    request.customer_id = customer_id
+    request.conversions.extend(conversions)
+    request.partial_failure = True
+
+    try:
+      conversion_upload_response = self._conversion_upload_service.upload_click_conversions(
+        request=request,
+      )
+    except GoogleAdsException as error:
+      print(f"Caught GoogleAdsException: {error}")
+      raise
+
+    for i, result in enumerate(conversion_upload_response.results):
+      print(f"Conversion {i}: {result}")
+
+    return self._get_partial_failures(conversion_upload_response)
+
+  def _get_partial_failures(self, response: Any) -> PartialFailures:
+    """Checks whether a response message has a partial failure error.
+
+    In Python the partial_failure_error attr is always present on a response
+    message and is represented by a google.rpc.Status message. So we can't
+    simply check whether the field is present, we must check that the code is
+    non-zero. Error codes are represented by the google.rpc.Code proto Enum:
+    https://github.com/googleapis/googleapis/blob/master/google/rpc/code.proto
+
+    Args:
+        response:  A MutateAdGroupsResponse message instance.
+
+    Returns: An empty dict if no partial failures exist, or a dict of the index
+      index mapped to the error message.
+    """
+    partial_failure = getattr(response, "partial_failure_error", None)
+    code = getattr(partial_failure, "code", None)
+    if code == 0:
+      # No failures.
+      print("No partial failures found.")
+      return {}
+
+    error_details = getattr(partial_failure, "details", [])
+
+    partial_failures = defaultdict(str)
+
+    for error_detail in error_details:
+      # Retrieve an instance of the GoogleAdsFailure class from the client
+      failure_message = self._client.get_type("GoogleAdsFailure")
+      # Parse the string into a GoogleAdsFailure message instance.
+      # To access class-only methods on the message we retrieve its type.
+      GoogleAdsFailure = type(failure_message)
+      failure_object = GoogleAdsFailure.deserialize(error_detail.value)
+
+      for error in failure_object.errors:
+        index = error.location.field_path_elements[0].index
+        message = f'Code: {error.error_code}, Error: {error.message}'
+        partial_failures[index] += message  # Can be multiple errors for the same conversion.
+
+    print(f"Partial failures: {partial_failures}")
+
+    return partial_failures
+
+  @staticmethod
+  def schema() -> Optional[ProtocolSchema]:
```

Comment:
The schema function should contain the fields that are expected to be present at the `config` object. I think the ones here refer to fields in the underlying customer data when it should really have the fields that you have under `_REQUIRED_GOOGLE_ADS_CREDENTIALS` and anything else you expect to be in the config object.

Reply:
Done. Ohhh, I see. Confusion on my part.